File: src/main/java/cz/astro/var/data/SpringConfig.java
package cz.astro.var.data;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.module.kotlin.KotlinModule;
import cz.astro.var.data.czev.controller.DownloadableTextMessageConverter;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.autoconfigure.http.HttpMessageConverters;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;
import org.springframework.data.transaction.ChainedTransactionManager;
import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.web.client.RestTemplate;

@Configuration
public class SpringConfig {

    @Bean
    @Primary
    public ChainedTransactionManager chainedTransactionManager(
            @Qualifier("czevTransactionManager") PlatformTransactionManager czevTransactionManager,
            @Qualifier("ocTransactionManager") PlatformTransactionManager ocTransactionManager
    ) {
        return new ChainedTransactionManager(ocTransactionManager, czevTransactionManager);
    }

    @Bean
    public RestTemplate restTemplate() {
        return new RestTemplate();
    }

    @Bean
    public HttpMessageConverters downloadableTextMessageConverter() {
        return new HttpMessageConverters(new DownloadableTextMessageConverter());
    }

    @Bean
    public Jackson2ObjectMapperBuilder objectMapperBuilder() {
        return new Jackson2ObjectMapperBuilder().modulesToInstall(new KotlinModule());
    }

    @Bean
    public MappingJackson2HttpMessageConverter mappingJackson2HttpMessageConverter()
    {
        ObjectMapper mapper = new ObjectMapper().registerModule(new KotlinModule());
        return new MappingJackson2HttpMessageConverter(mapper);
    }

}

File: backend/src/main/kotlin/cz/astro/var/data/SpringConfig.kt
package cz.astro.`var`.data

import cz.astro.`var`.data.czev.controller.DownloadableTextMessageConverter
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.boot.autoconfigure.http.HttpMessageConverters
import org.springframework.cache.annotation.EnableCaching
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.data.transaction.ChainedTransactionManager
import org.springframework.format.FormatterRegistry
import org.springframework.scheduling.annotation.EnableAsync
import org.springframework.transaction.PlatformTransactionManager
import org.springframework.web.servlet.config.annotation.CorsRegistry
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer

@Configuration
@EnableAsync
@EnableCaching
class SpringConfig {

    @Bean
    @Primary
    fun chainedTransactionManager(
            @Qualifier("czevTransactionManager") czevTransactionManager: PlatformTransactionManager,
            @Qualifier("ocTransactionManager") ocTransactionManager: PlatformTransactionManager
    ): ChainedTransactionManager {
        return ChainedTransactionManager(ocTransactionManager, czevTransactionManager)
    }

    @Bean
    fun downloadableTextMessageConverter(): HttpMessageConverters {
        return HttpMessageConverters(DownloadableTextMessageConverter())
    }

    @Bean
    fun webMvcConfigurer(): WebMvcConfigurer {
        return object : WebMvcConfigurer {
            override fun addFormatters(registry: FormatterRegistry?) {}

            override fun addCorsMappings(registry: CorsRegistry?) {
                registry!!.addMapping("/api/**")
                        .allowedOrigins("http://localhost:3000")
                        .allowedMethods("*")
                        .allowedHeaders("*")
                        .allowCredentials(true)
                        .maxAge(3600)
            }
        }
    }

}
